### ISSUE TITLE
False positive for "Statement could be decoupled from foreach"

### ISSUE THREAD

**Issue by zerkms** [kalessil/phpinspectionsea#1648]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) v4.0.6 |
| Language level | PHP 7.2                   |

Current behaviour

```php
        foreach ($copy as $key) {
            $s3From->getObject([
                'Bucket' => $bucket,
                'Key' => $key,
                'SaveAs' => $tmpPath,
            ]);

            // ... some more irrelevant lines of code

            unlink($tmpPath);
        }
```

The line with `unlink` is marked as could be decoupled.

I think as long as a variable is touched in any way (not only mutating, but just reading as well) - then it should not be subject of such analysis.

Expected behaviour
--

It should not report, because that statement is not a loop invariant.

Environment details
---

**Comment by orklah** [comment]
Hard to say based on your example alone. Without context, I'm also deducing that you're trying to delete a file based on a path that doesn't change. So if your loop does 100 iterations, you're trying to delete 100 times the same file and thus, the unlink should be out of the loop.

If the getObject actually create the file, it should also have a method to delete it to stay coherent.

There would be no point to consider variable reading as coupled, might as well delete the inspection...

**Comment by zerkms** [comment]
> Hard to say based on your example alone. Without context, I'm also deducing that you're trying to delete a file based on a path that doesn't change.

The first loop statement has `'SaveAs' => $tmpPath,` - it's where the file is downloaded to. Every iteration a new file is downloaded to the same path.

**Comment by orklah** [comment]
Nope, not familiar with it, I was talking more generally. It's weird that a method create a file and let you deal with the suppression yourself if it's designed to be temporary.

What I meant about deletion is that if you fix the inspection not to report variable reading, there will be no code that fits the criteria.

Any method can't get decoupled
Any variable, read or write can't get decoupled
Any echo can't get decoupled

There would me nothing left to report :)
